### ISSUE TITLE
[svg-native] Support for XML entities/CDATA required?

### ISSUE THREAD

**Issue by dirkschulze** [w3c/svgwg#672]
Could remove the requirement to support XML entities? Maybe even say implementations _should_ not support it? XML entities require a special XML parser. Furthermore, there are attack patterns like the [billion laughs attack](https://en.wikipedia.org/wiki/Billion_laughs_attack) that could mean a security but at least reliability threat.

SVG OT uses _should not_ as well.

**Comment by litherum** [comment]
That’s interesting. I was under the impression that ::not:: having XML entities required a special XML parser.

**Comment by dirkschulze** [comment]
I meant to write an _actual_ XML parser :P

**Comment by litherum** [comment]
Given that SVG is an XML dialect, using an actual XML parser seems like a good idea. I expect most implementations use out-of-the-box parsers like [NSXMLParser](https://developer.apple.com/documentation/foundation/nsxmlparser?language=objc) on macOS and iOS, [Windows.Data.XML.DOM](https://docs.microsoft.com/en-us/uwp/api/windows.data.xml.dom) on Windows, [libxml2](http://xmlsoft.org/) on Linux, and [XMLPullParser](https://developer.android.com/reference/org/xmlpull/v1/XmlPullParser.html) on Android. Requiring an implementation to write their own parser would be unfortunate.

**Comment by dirkschulze** [comment]
@litherum I don’t think that we can make such an assumption. Note: it is not about disallowing XML entities but about not requiring support for XML entities.

**Comment by svgeesus** [comment]
XML entities (and all the rest of the internal DTD subset) is required by a fully conforming XML parser. Note that this is entirely separate from whether it uses the DTD (internal subset, and external subset) for DTD validation. Even a parser which just does well-formed checks, needs to parse *and then ignore* all the DTD syntax, although it is still required to perform entity substitution and fill in attribute defaults.

This is a well-known problem, and means that most of the implementation burden for an XML parser is for parts of the language that are only used in DTDs.

It also is a [security issue](https://en.wikipedia.org/wiki/XML_external_entity_attack), because it is trivially easy to make memory explore. Entity A is "1234567890" and entity B is ten of entity A, and entity Z requires petabytes of storage.

This has caused some to [call for an XML 2.0](https://norman.walsh.name/2008/02/20/xml20) or to [propose MicroXML](https://www.w3.org/community/microxml/wiki/Main_Page) which just has the document parts of XML and not the whole DTD part. (A major blocker to that is the use by HTML5 of a mininal DOCTYPE as a flag).

Regardless of standardization efforts in that space, in my view SVG Native should

a) require only the DTD-less subset of XML be supported (so a parser that handles only the document subset is fully conforming)
b) make the use of DOCTYPE be non-conforming for SVG Native content

**Comment by grorg** [comment]
@svgeesus 

> a) require only the DTD-less subset of XML be supported (so a parser that handles only the document subset is fully conforming)

So no XML entities? I agree.

What should happen if a document does provide a DTD?

> b) make the use of DOCTYPE be non-conforming for SVG Native content

What should an implementation do if it sees one? Even if it is something like:

```html
<!doctype html>
<svg ....
```

Last question, did you have thoughts on CDATA?

**Comment by litherum** [comment]
There are competing goals:

1. Simplicity. From a systems perspective, not having to perform entity expansion or CDATA parsing is undeniably simpler.
2. Implementability. It should be possible to implement using an off-the-shelf XML parser without spending a ton of effort configuring it for exactly our situation.
3. Conformity. Allowing too many "conforming" implementations to have wildly different behavior means we, as a standardization body, have failed.

We're missing some data on 2. I'll do some research to determine what some off-the-shelf parsers do by default and how difficult it is to configure them to have the opposite behavior.

**Comment by dirkschulze** [comment]
@litherum I do **not** want to forbid engines to support DTD/Entities or doctypes. I just don’t think it should be a **requirement** and **documents** _should_ avoid it.

**Comment by AmeliaBR** [comment]
To clarify: we are talking here about the ability of an XML document to declare custom entities?

Adobe software uses these (or used to) for declaring custom namespaces & such. I've had to hand-edit them out of files before running them through SVGO, because the XML parser it used didn't recognize them.

I definitely agree with Dirk's general rule: we should never forbid software from supporting the full XML syntax. But it might make sense to define that certain parts of XML aren't required (just like we're saying that certain parts of CSS & SVG aren't required), and that conforming documents therefore should not use them.

**Comment by svgeesus** [comment]
Precedent from [SOAP](https://www.w3.org/TR/2007/REC-soap12-part1-20070427/#soapenv):

> The XML infoset of a SOAP message MUST NOT contain a document type declaration information item.

**Comment by svgeesus** [comment]
> To clarify: we are talking here about the ability of an XML document to declare custom entities?

Custom entities, default attributes values, new elements, new attributes.

**Comment by svgeesus** [comment]
Looking at [RAPIDXML Manual](http://rapidxml.sourceforge.net/manual.html#namespacerapidxml_1dependencies_and_compatibility) I see:

> Rapidxml understands and expands the following character references: &apos; &amp; &quot; &lt; &gt; &#...; Other character references are not expanded.

and even more interesting

> RapidXml is not a W3C compliant parser, primarily because it ignores DOCTYPE declarations.

**Comment by litherum** [comment]
From some initial testing, it looks like Microsoft's and Apple's implementations expand entities automatically, and Adobe's implementation doesn't.

**Comment by litherum** [comment]
Also, from some testing, it looks like, by default, libxml2 and NSXMLParser expand entities, while by default, Windows.Data.XML.Dom and RapidXML don't.

**Comment by svgeesus** [comment]
All of which seems to indicate that, for best interoperability, a valid SVG Native document must not contain any of the external or internal XML DTD subset (and thus, **define** no entities). That part we already resolved on.

It also implies that a valid SVG Native document must not **use** entities. Which is fine because, compared to 1998 when XML and SVG were starting out, Unicode is widespread and if you want é or Ť or ɔʇə you just type them or pick them, rather than relying on character entities like `&eacute;` and so on.

**Comment by tabatkins** [comment]
Beyond the &lt;/&gt;, &amp;, and &apos;/&quot;, right?

If so, yes I agree. The need for entities beyond those five no longer exists.

**Comment by AmeliaBR** [comment]
Yes, we definitely need to support the basic entities defined in XML. In addition to the ones Tab mentioned, I'd suggest including [character references based on codepoint values](https://www.w3.org/TR/2008/REC-xml-20081126/#sec-references) for tool chains that aren't Unicode friendly. I'm pretty sure any XML tool supports that much.

Beyond that, there would be no way to create a valid named entity, so using any other entity would be equally invalid.
